File: android-app/app/src/main/java/com/chainlesschain/android/remote/commands/SecurityCommands.kt
package com.chainlesschain.android.remote.commands

import com.chainlesschain.android.remote.client.RemoteCommandClient
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

/**
 * 安全操作命令 API
 *
 * 提供类型安全的安全相关命令
 */
@Singleton
class SecurityCommands @Inject constructor(
    private val client: RemoteCommandClient
) {
    /**
     * 锁定工作站
     */
    suspend fun lockWorkstation(): Result<LockResponse> {
        return client.invoke("security.lockWorkstation", emptyMap())
    }

    /**
     * 获取安全状态摘要
     */
    suspend fun getStatus(): Result<SecurityStatusResponse> {
        return client.invoke("security.getStatus", emptyMap())
    }

    /**
     * 获取活动用户
     */
    suspend fun getActiveUsers(): Result<ActiveUsersResponse> {
        return client.invoke("security.getActiveUsers", emptyMap())
    }

    /**
     * 获取登录历史
     *
     * @param limit 返回数量限制
     */
    suspend fun getLoginHistory(limit: Int = 50): Result<LoginHistoryResponse> {
        val params = mapOf("limit" to limit)
        return client.invoke("security.getLoginHistory", params)
    }

    /**
     * 获取防火墙状态
     */
    suspend fun getFirewallStatus(): Result<FirewallStatusResponse> {
        return client.invoke("security.getFirewallStatus", emptyMap())
    }

    /**
     * 获取杀毒软件状态
     */
    suspend fun getAntivirusStatus(): Result<AntivirusStatusResponse> {
        return client.invoke("security.getAntivirusStatus", emptyMap())
    }

    /**
     * 获取加密状态
     */
    suspend fun getEncryptionStatus(): Result<EncryptionStatusResponse> {
        return client.invoke("security.getEncryptionStatus", emptyMap())
    }

    /**
     * 获取系统更新状态
     */
    suspend fun getUpdates(): Result<UpdatesStatusResponse> {
        return client.invoke("security.getUpdates", emptyMap())
    }
}

// 响应数据类

@Serializable
data class LockResponse(
    val success: Boolean,
    val message: String
)

@Serializable
data class SecurityStatusResponse(
    val success: Boolean,
    val security: SecuritySummary
)

@Serializable
data class SecuritySummary(
    val firewallEnabled: Boolean? = null,
    val antivirusInstalled: Boolean? = null,
    val encryptionEnabled: Boolean? = null,
    val pendingUpdates: Int? = null,
    val platform: String
)

@Serializable
data class ActiveUsersResponse(
    val success: Boolean,
    val users: List<ActiveUser>,
    val count: Int,
    val currentUser: String
)

@Serializable
data class ActiveUser(
    val username: String,
    val domain: String? = null,
    val terminal: String? = null,
    val loginTime: String? = null,
    val logonType: String? = null
)

@Serializable
data class LoginHistoryResponse(
    val success: Boolean,
    val history: List<LoginRecord>,
    val count: Int
)

@Serializable
data class LoginRecord(
    val username: String? = null,
    val terminal: String? = null,
    val time: String? = null,
    val type: String? = null,
    val details: String? = null
)

@Serializable
data class FirewallStatusResponse(
    val success: Boolean,
    val enabled: Boolean? = null,
    val type: String? = null,
    val profiles: List<FirewallProfile>? = null,
    val error: String? = null,
    val defaultInbound: String? = null,
    val defaultOutbound: String? = null,
    val ruleCount: Int? = null,
    val profile: String? = null,
    val rules: Int? = null
)

@Serializable
data class FirewallProfile(
    val name: String,
    val enabled: Boolean
)

@Serializable
data class AntivirusStatusResponse(
    val success: Boolean,
    val installed: Boolean? = null,
    val products: List<AntivirusProduct>? = null,
    val error: String? = null
)

@Serializable
data class AntivirusProduct(
    val name: String,
    val state: Int? = null,
    val builtin: Boolean? = null
)

@Serializable
data class EncryptionStatusResponse(
    val success: Boolean,
    val enabled: Boolean? = null,
    val type: String? = null,
    val percentage: Int? = null,
    val error: String? = null,
    val driveName: String? = null,
    val encrypted: Boolean? = null,
    val locked: Boolean? = null,
    val algorithm: String? = null,
    val method: String? = null,
    val status: String? = null
)

@Serializable
data class UpdatesStatusResponse(
    val success: Boolean,
    val pendingCount: Int? = null,
    val updates: List<PendingUpdate>? = null,
    val error: String? = null
)

@Serializable
data class PendingUpdate(
    val title: String,
    val kb: String? = null
)

File: ios-app/Modules/CoreP2P/Sources/RemoteSkills/Security/SecurityCommands.swift
import Foundation

/// 安全状态查询 typed RPC wrapper — Phase 6.1B1。
///
/// 镜像 Android `app/src/main/java/.../remote/commands/SecurityCommands.kt`。
/// 8 method 全 wired，桌面 `security-handler.js` 与 Android 100% 名称对齐
/// （Coverage doc §1.4：A=8 D=8 ✓=8）。
///
/// **wire 协议**（与桌面 `security-handler.js` 对齐）：
/// - `security.lockWorkstation` — 锁屏（mutating；高危需 ApprovalGate）
/// - `security.getStatus` — 安全状态摘要（防火墙/AV/加密/更新四合一）
/// - `security.getActiveUsers` / `security.getLoginHistory` — 用户/登录历史
/// - `security.getFirewallStatus` / `security.getAntivirusStatus` /
///   `security.getEncryptionStatus` / `security.getUpdates` — 单项详情
public actor SecurityCommands {

    private let client: RemoteCommandClient

    public init(client: RemoteCommandClient) {
        self.client = client
    }

    /// 锁屏（mutating；桌面端高危，应配 ApprovalGate UI）。
    public func lockWorkstation(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> LockResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "security.lockWorkstation",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, LockResponse.decode)
    }

    /// 安全状态摘要（4 字段：防火墙 / AV / 加密 / 待更新数）。
    public func getStatus(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> SecurityStatusResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "security.getStatus",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, SecurityStatusResponse.decode)
    }

    /// 当前活动用户列表。
    public func getActiveUsers(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> ActiveUsersResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "security.getActiveUsers",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, ActiveUsersResponse.decode)
    }

    /// 登录历史（默认最近 50 条）。
    public func getLoginHistory(
        pcPeerId: String,
        limit: Int = 50,
        mobileDid: String? = nil
    ) async throws -> LoginHistoryResponse {
        guard limit > 0 else {
            throw RemoteSkillError.invalidArgument("security.getLoginHistory: limit must be > 0")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "security.getLoginHistory",
            params: ["limit": limit],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, LoginHistoryResponse.decode)
    }

    /// 防火墙状态（含 profiles 列表 + ruleCount）。
    public func getFirewallStatus(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> FirewallStatusResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "security.getFirewallStatus",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, FirewallStatusResponse.decode)
    }

    /// 杀毒软件状态（installed + products 列表）。
    public func getAntivirusStatus(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> AntivirusStatusResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "security.getAntivirusStatus",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, AntivirusStatusResponse.decode)
    }

    /// 加密状态（BitLocker / FileVault / LUKS 等）。
    public func getEncryptionStatus(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> EncryptionStatusResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "security.getEncryptionStatus",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, EncryptionStatusResponse.decode)
    }

    /// 待安装系统更新。
    public func getUpdates(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> UpdatesStatusResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "security.getUpdates",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, UpdatesStatusResponse.decode)
    }

    private static func decode<T>(
        _ resp: TerminalRpcResponse,
        _ decoder: (String) throws -> T
    ) throws -> T {
        switch resp {
        case .success(_, let resultJson):
            return try decoder(resultJson)
        case .failure(let reqId, let msg):
            throw RemoteSkillError.remoteError(reqId: reqId, message: msg)
        }
    }
}
